File: workflows/java/sdk/order-processor/src/main/java/io/dapr/quickstarts/workflows/activities/RequestApprovalActivity.java
package io.dapr.quickstarts.workflows.activities;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.dapr.quickstarts.workflows.models.OrderPayload;
import io.dapr.workflows.WorkflowActivity;
import io.dapr.workflows.WorkflowActivityContext;

public class RequestApprovalActivity implements WorkflowActivity {
  private static Logger logger = LoggerFactory.getLogger(RequestApprovalActivity.class);

  @Override
  public Object run(WorkflowActivityContext ctx) {
    OrderPayload order = ctx.getInput(OrderPayload.class);
    logger.info("Requesting approval for order: {}", order);

    // Simulate slow processing & sending the approval to the recipient
    try {
      Thread.sleep(2 * 1000);
    } catch (InterruptedException e) {
    }

    return "";
  }

}

File: workflows/csharp/sdk/order-processor/Activities/RequestApprovalActivity.cs
namespace WorkflowConsoleApp.Activities;

using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Dapr.Workflow;
using WorkflowConsoleApp.Models;

internal sealed partial class RequestApprovalActivity(ILogger<RequestApprovalActivity> logger) : WorkflowActivity<ApprovalRequest, object?>
{
    public override async Task<object?> RunAsync(WorkflowActivityContext context, ApprovalRequest approvalRequest)
    {
        LogRequestApproval(logger, approvalRequest);

        // Simulate slow processing & sending the approval to the recipient
        await Task.Delay(TimeSpan.FromSeconds(2));

        return Task.FromResult<object?>(null);
    }
    
    [LoggerMessage(LogLevel.Information, "Approval Request {approvalRequest}")]
    static partial void LogRequestApproval(ILogger logger, ApprovalRequest approvalRequest);
}